File: output/cpp/tests/heavy_native_prolog.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>
#include <map>

// Test: Heavy native prolog and epilog with edge-case C++ syntax.

class NativeHelper {
    // Mentions @@system and -> $State in comments
    // These should NOT be parsed as Frame
public:
    std::string process(const std::map<std::string, std::string>& data) {
        // Comment with -> $Transition and => $^ and push$
        return std::string("{\"key\": \"value with { braces }\"}");
    }
};

// Global with @@ in string
std::map<std::string, std::string> GLOBAL_CONFIG = {
    {"description", "Config with @@system-like syntax"},
    {"pattern", "-> $Start"},
};

class HeavyNativePrologFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    HeavyNativePrologFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class HeavyNativePrologFrameContext {
public:
    HeavyNativePrologFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    HeavyNativePrologFrameContext(HeavyNativePrologFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class HeavyNativePrologCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<HeavyNativePrologFrameEvent> forward_event;
    std::unique_ptr<HeavyNativePrologCompartment> parent_compartment;

    explicit HeavyNativePrologCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<HeavyNativePrologCompartment> clone() const {
        auto c = std::make_unique<HeavyNativePrologCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class HeavyNativeProlog {
private:
    std::vector<std::unique_ptr<HeavyNativePrologCompartment>> _state_stack;
    std::unique_ptr<HeavyNativePrologCompartment> __compartment;
    std::unique_ptr<HeavyNativePrologCompartment> __next_compartment;
    std::vector<HeavyNativePrologFrameContext> _context_stack;

    void __kernel(HeavyNativePrologFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            HeavyNativePrologFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                HeavyNativePrologFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    HeavyNativePrologFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(HeavyNativePrologFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Start") {
            _state_Start(__e);
        }
    }

    void __transition(std::unique_ptr<HeavyNativePrologCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Start(HeavyNativePrologFrameEvent& __e) {
        if (__e._message == "run_test") {
            NativeHelper h;
            std::string result = h.process(GLOBAL_CONFIG);
            if (result.length() > 0) {
                printf("PASS: Heavy native prolog handled correctly\n");
            } else {
                assert(false);
            }
        }
    }

public:
    HeavyNativeProlog() {
        __compartment = std::make_unique<HeavyNativePrologCompartment>("Start");
        HeavyNativePrologFrameEvent __frame_event("$>");
        HeavyNativePrologFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void run_test() {
        HeavyNativePrologFrameEvent __e("run_test");
        HeavyNativePrologFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

// Heavy epilog
class PostSystem {
public:
    std::map<std::string, std::string> data = {{"@@system", "in dict"}, {"-> $State", "in dict"}};

    bool run() {
        // -> $NotATransition
        // => $^
        return true;
    }
};

int main() {
    HeavyNativeProlog s;
    s.run_test();
    PostSystem p;
    if (!p.run()) { assert(false); }
    return 0;
}

File: output/javascript/tests/heavy_native_prolog.js

// Test: Heavy native prolog and epilog with edge-case JavaScript syntax.

class NativeHelper {
    // Mentions @@system and -> $State in comments
    // These should NOT be parsed as Frame

    static TEMPLATE = `
    {
        "@@target": "not_real",
        "-> $State": "just a string",
        "push$": true
    }
    `;

    process(data) {
        const result = JSON.stringify({"key": "value with { braces }"});
        // Comment with -> $Transition and => $^ and push$
        return result;
    }
}

// Global with @@ in string
const GLOBAL_CONFIG = {
    "description": "Config with @@system-like syntax",
    "pattern": "-> $Start",
};

export class HeavyNativePrologFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class HeavyNativePrologFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class HeavyNativePrologCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new HeavyNativePrologCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class HeavyNativeProlog {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new HeavyNativePrologCompartment("Start");
        this.__next_compartment = null;
        const __frame_event = new HeavyNativePrologFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new HeavyNativePrologFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new HeavyNativePrologFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new HeavyNativePrologFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    run_test() {
        const __e = new HeavyNativePrologFrameEvent("run_test", null);
        const __ctx = new HeavyNativePrologFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    _state_Start(__e) {
        if (__e._message === "run_test") {
            const h = new NativeHelper();
            const result = h.process(GLOBAL_CONFIG);
            if (result !== null) {
                console.log("PASS: Heavy native prolog handled correctly");
            } else {
                throw new Error("FAIL");
            }
        }
    }
}

// Heavy epilog
class PostSystem {
    constructor() {
        this.data = {"@@system": "in dict", "-> $State": "in dict"};
    }

    run() {
        // -> $NotATransition
        // => $^
        return true;
    }
}

const s = new HeavyNativeProlog();
s.run_test();
const p = new PostSystem();
if (!p.run()) { throw new Error("epilog failed"); }
